src/methods.rs

```diff
@@ -92,6 +99,64 @@ impl LateLintPass for MethodsPass {
+                    if inner_name.node.as_str() == "map"
+                            && match_type(cx, cx.tcx.expr_ty(&inner_args[0]), &OPTION_PATH) {
+                        // lint message
+                        let msg =
+                            "called `map(f).unwrap_or(a)` on an Option value. This can be done \
+                             more directly by calling `map_or(a, f)` instead";
+                        // get args to map() and unwrap_or()
+                        let map_arg = snippet(cx, inner_args[1].span, "..");
+                        let unwrap_arg = snippet(cx, args[1].span, "..");
+                        // lint, with note if neither arg is > 1 line and both map() and
+                        // unwrap_or() have the same span
+                        if map_arg.lines().count() > 1 || unwrap_arg.lines().count() > 1
+                                || inner_args[1].span.expn_id != args[1].span.expn_id {
+                            span_lint(cx, OPTION_MAP_UNWRAP_OR, expr.span, msg);
+                        }
+                        else {
+                            let span = mk_sp(inner_name.span.lo, name.span.hi);
+                            span_note_and_lint(
+                                cx, OPTION_MAP_UNWRAP_OR, expr.span, msg, span,
+                                &format!("replace this with map_or({1}, {0})",
+                                         map_arg, unwrap_arg)
+                            );
+                        };
+                    }
+                }
+            }
+            // check Option.map(_).unwrap_or_else(_)
+            else if name.node.as_str() == "unwrap_or_else" {
+                if let ExprMethodCall(ref inner_name, _, ref inner_args) = args[0].node {
+                    if inner_name.node.as_str() == "map"
+                            && match_type(cx, cx.tcx.expr_ty(&inner_args[0]), &OPTION_PATH) {
+                        // lint message
+                        let msg =
+                            "called `map(f).unwrap_or_else(g)` on an Option value. This can be \
+                             done more directly by calling `map_or_else(g, f)` instead";
+                        // get args to map() and unwrap_or_else()
+                        let map_arg = snippet(cx, inner_args[1].span, "..");
+                        let unwrap_arg = snippet(cx, args[1].span, "..");
+                        // lint, with note if neither arg is > 1 line and both map() and
+                        // unwrap_or_else() have the same span
+                        if map_arg.lines().count() > 1 || unwrap_arg.lines().count() > 1
+                                || inner_args[1].span.expn_id != args[1].span.expn_id {
+                            span_lint(cx, OPTION_MAP_UNWRAP_OR_ELSE, expr.span, msg);
+                        }
+                        else {
+                            let span = mk_sp(inner_name.span.lo, name.span.hi);
```

Comment:
I was talking about this issue on #rust-internals, and some people brought up the possibility that the call to `map()` might come from a macro expansion, and thus might not even be located in the same file as the call to `unwrap_or()`/`unwrap_or_else()`. For example, you might have `some_macro!(foo).unwrap_or(bar)`, where `some_macro!()` is defined in another file. This would keep the note from displaying properly. However, now that I think about it a little more carefully, we probably wouldn't want to lint in that case in the first place. Should I not lint in the case that the spans are different, or should I simply remove that check altogether?